h2o-genmodel/src/main/java/hex/genmodel/algos/targetencoder/TargetEncoderMojoModel.java

```diff
@@ -0,0 +1,123 @@
+package hex.genmodel.algos.targetencoder;
+
+import hex.genmodel.MojoModel;
+
+import java.util.*;
+
+public class TargetEncoderMojoModel extends MojoModel {
+
+  public TargetEncoderMojoModel(String[] columns, String[][] domains, String responseName) {
+    super(columns, domains, responseName);
+  }
+
+  public EncodingMaps _targetEncodingMap;
+  public Map<String, Integer> _teColumnNameToIdx = new HashMap<>();
+  public boolean _withBlending;
+  public double _inflectionPoint;
+  public double _smoothing;
+
+  // Could be passed from the model training phase so that we don't have to recompute it here. Or maybe it is fine as it should be computed only once per mojoModel
+  private double _priorMean = -1;
+
+  public static double computeLambda(int nrows, double inflectionPoint, double smoothing) {
+    return 1.0 / (1 + Math.exp((inflectionPoint - nrows) / smoothing));
+  }
+  
+  public static double computeBlendedEncoding(double lambda, double posteriorMean, double priorMean) {
+    return lambda * posteriorMean + (1 - lambda) * priorMean;
+  }
+  
+  /**
+   * Computes prior mean i.e. unconditional mean of the response. Should be the same for all columns
+   */
+  public static double computePriorMean(EncodingMap encodingMap) {
+    int sumOfNumerators = 0;
+    int sumOfDenominators = 0;
+    Iterator<int[]> iterator = encodingMap._encodingMap.values().iterator();
+    while( iterator.hasNext()) {
+      int[] next = iterator.next();
+      sumOfNumerators += next[0];
+      sumOfDenominators += next[1];
+    }
+    return (double) sumOfNumerators / sumOfDenominators;
+  }
+  
+  @Override
+  public double[] score0(double[] row, double[] preds) {
+    if(_targetEncodingMap != null) {
+      
+      int predictionIndex = 0;
+      // Order of entries is not defined in sets(hash map). So we need some source of consistency.
+      // Following will guarantee order of transformations. Ascending order based on index of te column in the input
+      LinkedHashMap<String, EncodingMap> sortedByColumnIndex = sortByColumnIndex(_targetEncodingMap.encodingMap());
+
+      for (Map.Entry<String, EncodingMap> columnToEncodingsMap : sortedByColumnIndex.entrySet() ) {
+        
+        String teColumn = columnToEncodingsMap.getKey();
+        int indexOfColumnInRow = _teColumnNameToIdx.get(teColumn);
+        
+        int originalValue = Double.isNaN(row[indexOfColumnInRow]) ? -1 :  (int)row[indexOfColumnInRow]; // original categorical level represented as index
+        EncodingMap encodings = columnToEncodingsMap.getValue();
+
+        _priorMean = _priorMean == -1 ? computePriorMean(encodings) : _priorMean;
```

Comment:
What if the actual mean is `-1` - are we going to recalculate it every time?

Reply:
@michalkurka Please take a look at https://github.com/h2oai/h2o-3/pull/3282#discussion_r299887028 answer.